### PR TITLE
Added validation for project Name

### PR DESCRIPTION
**Pull request by hiteshwani29**
### What does this PR change?

- This PR will add project name validation in the create project API. This PR will fix the issue #148 and https://github.com/paralus/cli/issues/46

### Does the PR depend on any other PRs or Issues? If yes, please list them.

-

### Checklist

I confirm, that I have...

- [x] Read and followed the contributing guide in `CONTRIBUTING.md`
- [ ] Added tests for this PR
- [x] Formatted the code using `go fmt` (if applicable)
- [ ] Updated [documentation on the Paralus docs site](https://github.com/paralus/website/blob/main/docs) (if applicable)
- [ ] Updated `CHANGELOG.md`

### CODE REVIEW

**Comment by hiteshwani29** [comment]
please review this PR
@niravparikh05 @akshay196 @mabhi

File: pkg/common/constants.go
Comment (niravparikh05) [inline pkg/common/constants.go:55]: this regex looks good, except that name like `my-$-project` is allowed which inturn gets replaced and created with the name my-himBHsproject which we should not be allowing.

let the valid characters be same as given regex and hyphen, except at the beginning or the end. No other special chars allowed.

File: pkg/service/project.go
Comment (niravparikh05) [inline pkg/service/project.go:66]: can the error message be more user friendly ? It would not make showing a regex expression in the error message for non tech related users. 

It can be something like "Valid characters are a-z0-9 and hyphen, except at the beginning or the end"

**Review by niravparikh05 — CHANGES_REQUESTED** [review]
@hiteshwani29 thanks for your pr, really appreciate your contribution towards paralus. There are a few minor comments, otherwise this looks good.

Also can you ensure you run go fmt before commiting your changes as gh actions are currently failing due to it. 

Ensure to update changelog.md as well, you can follow the same format as in the file.

Let me know if you have any doubts.

**Comment by niravparikh05** [comment]
@hiteshwani29 Issue #149 is similar to this, it can use the same regex validations during cluster creation ? Will it be possible for you to make changes for the same with this pr ?

File: pkg/service/project.go
Comment (akshay196) [reply pkg/service/project.go:66]: Or `Valid characters are alphanumeric and hyphen, except at the beginning or the end`

File: pkg/service/utils.go
Comment (akshay196) [inline pkg/service/utils.go:40]: This can be eliminated by doing build time regex compilation.

```
var PrjNameRX = regexp.MustCompile(`^[a-zA-Z]([-a-zA-Z0-9]*[a-zA-Z0-9])?$`)
```

File: pkg/common/constants.go
Comment (akshay196) [reply pkg/common/constants.go:55]: @hiteshwani29 Any reason why all letters except first are grouped and added `()?` quantifier? Can this just be `^[a-zA-Z][-a-zA-Z0-9]*[a-zA-Z0-9]$`

Comment (hiteshwani29) [reply pkg/common/constants.go:55]: No @akshay196 not because of any specific reason I will make the changes in regex respectively .

File: pkg/service/project.go
Comment (hiteshwani29) [reply pkg/service/project.go:66]: Agree will make the message more user-friendly.

File: pkg/common/constants.go
Comment (hiteshwani29) [reply pkg/common/constants.go:55]: @niravparikh05 
Not sure but the cobra library is not parsing string arguments correctly with special characters such as '$' or '$$'. 
- for `pctl creat project my-$-project ` It replaces the $ with some string char and gives us some incorrect name in args array.
- but if we pass the project name with escape char `my-\$-project` or with a single quote `'my-$-project'` in the command. It parses the arg correctly in the args array and throws a validating error

**Comment by hiteshwani29** [comment]
Hey @niravparikh05 @akshay196 


> @hiteshwani29 Issue #149 is similar to this, it can use the same regex validations during cluster creation ? Will it be possible for you to make changes for the same with this pr ?

@niravparikh05 we already have the  validation for the cluster name
https://github.com/paralus/paralus/blob/main/pkg/service/cluster.go#L124

**Comment by hiteshwani29** [comment]
@niravparikh05 @akshay196 
Please review the changes and let me know.

File: pkg/common/constants.go
Comment (akshay196) [inline pkg/common/constants.go:55]: This can be replaced with regex directly:
```go
var PrjNameRX = regexp.MustCompile(`^[a-zA-Z][-a-zA-Z0-9]*[a-zA-Z0-9]$`)
```
and used as:
```go
PrjNameRX.MatchString(project.Metadata.GetName())
```

No need of `MatchStringToRegx` func.

**Comment by akshay196** [comment]
@hiteshwani29 One single change ⬆️ . Otherwise your code looks good to me.

File: pkg/common/constants.go
Comment (hiteshwani29) [reply pkg/common/constants.go]: Done @akshay196

**Review by akshay196 — APPROVED** [review]

**Comment by hiteshwani29** [comment]
@akshay196 I don't have merge access. Can you merge this PR and close this issue?

**Comment by akshay196** [comment]
Waiting for @niravparikh05 to appove.

**Comment by hiteshwani29** [comment]
> Waiting for @niravparikh05 to appove.

ok @akshay196. will wait for him.

**Comment by akshay196** [comment]
@niravparikh05 After your approval, squash commits and change message to something like `fix: add project name validation` before merging. let's try conventional commits to achieve https://github.com/paralus/paralus/issues/177 from here.

**Review by akshay196 — APPROVED** [review]

**Review by niravparikh05 — APPROVED** [review]
lgtm

**Comment by akshay196** [comment]
@hiteshwani29 Thanks for contributing and congratulation on your first PR. 🥳

**Comment by hiteshwani29** [comment]
Thanks 🙏 @akshay196 
Looking forward to make more contribution in paralus.

**Pull request merged**
